File: go/internal/agent/fixflow/tools.go
package fixflow

import (
	"os"
	"sort"

	"google.golang.org/adk/tool"
	"google.golang.org/adk/tool/functiontool"
)

type readFileArgs struct {
	Path string `json:"path"`
}
type readFileResult struct {
	Content string `json:"content"`
}
type listDirArgs struct {
	Path string `json:"path"`
}
type listDirResult struct {
	Entries []string `json:"entries"`
}

// repoTools returns read-only tools (read_file, list_dir) rooted at the checkout, so
// a tool-using agent can examine the real repository — its standards docs, existing
// tests, and layout — and ground decisions in what the repo actually does.
func repoTools(root string) ([]tool.Tool, error) {
	readFile, err := functiontool.New(functiontool.Config{
		Name:        "read_file",
		Description: "Read a repository file by its repo-relative path (e.g. \"src/main.go\" or \"AGENTS.md\").",
	}, func(_ tool.Context, args readFileArgs) (readFileResult, error) {
		c, err := ReadFile(root, args.Path)
		if err != nil {
			return readFileResult{}, err
		}
		return readFileResult{Content: c}, nil
	})
	if err != nil {
		return nil, err
	}

	listDir, err := functiontool.New(functiontool.Config{
		Name:        "list_dir",
		Description: "List the files and subdirectories of a repository directory by its repo-relative path. Use \".\" for the repository root.",
	}, func(_ tool.Context, args listDirArgs) (listDirResult, error) {
		entries, err := listDirEntries(root, args.Path)
		if err != nil {
			return listDirResult{}, err
		}
		return listDirResult{Entries: entries}, nil
	})
	if err != nil {
		return nil, err
	}

	return []tool.Tool{readFile, listDir}, nil
}

// listDirEntries lists a checkout directory (path-safe), suffixing subdirectories
// with "/" and hiding the .git directory.
func listDirEntries(root, rel string) ([]string, error) {
	full, err := safeJoin(root, rel)
	if err != nil {
		return nil, err
	}
	ents, err := os.ReadDir(full)
	if err != nil {
		return nil, err
	}
	names := make([]string, 0, len(ents))
	for _, e := range ents {
		if e.Name() == ".git" {
			continue
		}
		name := e.Name()
		if e.IsDir() {
			name += "/"
		}
		names = append(names, name)
	}
	sort.Strings(names)
	return names, nil
}

File: kotlin/src/main/kotlin/io/github/jkjamies/automationagent/agent/fixflow/Tools.kt
package io.github.jkjamies.automationagent.agent.fixflow

import com.google.adk.kt.tools.BaseTool
import com.google.adk.kt.tools.ToolContext
import com.google.adk.kt.types.FunctionDeclaration
import com.google.adk.kt.types.Schema
import com.google.adk.kt.types.Type

/**
 * Read-only tools (read_file, list_dir) rooted at the checkout, so a tool-using agent can examine
 * the real repository — its standards docs, existing tests, and layout — and ground decisions in
 * what the repo actually does. Tool failures self-wrap as `{"error": …}` rather than aborting the
 * run.
 */
internal fun repoTools(root: String): List<BaseTool> = listOf(ReadFileTool(root), ListDirTool(root))

private class ReadFileTool(private val root: String) :
    BaseTool(
        name = "read_file",
        description = "Read a repository file by its repo-relative path (e.g. \"src/main.kt\" or \"AGENTS.md\").",
    ) {
    override fun declaration(): FunctionDeclaration =
        FunctionDeclaration(
            name = name,
            description = description,
            parameters = Schema(type = Type.OBJECT, properties = mapOf("path" to Schema(type = Type.STRING, description = "repo-relative path")), required = listOf("path")),
        )

    override suspend fun run(context: ToolContext, args: Map<String, Any>): Any =
        try {
            mapOf("content" to readFile(root, args["path"] as? String ?: ""))
        } catch (e: Exception) {
            mapOf("error" to (e.message ?: e.toString()))
        }
}

private class ListDirTool(private val root: String) :
    BaseTool(
        name = "list_dir",
        description = "List the files and subdirectories of a repository directory by its repo-relative path. Use \".\" for the repository root.",
    ) {
    override fun declaration(): FunctionDeclaration =
        FunctionDeclaration(
            name = name,
            description = description,
            parameters = Schema(type = Type.OBJECT, properties = mapOf("path" to Schema(type = Type.STRING, description = "repo-relative path")), required = listOf("path")),
        )

    override suspend fun run(context: ToolContext, args: Map<String, Any>): Any =
        try {
            mapOf("entries" to listDirEntries(root, args["path"] as? String ?: "."))
        } catch (e: Exception) {
            mapOf("error" to (e.message ?: e.toString()))
        }
}
